### PR TITLE
refactor Layout

### PR DESCRIPTION
**Pull request by zilongshanren**

### CODE REVIEW

**Comment by zilongshanren** [comment]
please don't merge this PR, it still need to be improved.

**Comment by zilongshanren** [comment]
[ci rebuild]

**Comment by zilongshanren** [comment]
[ci rebuild]

**Comment by zilongshanren** [comment]
[ci rebuild]

**Comment by ricardoquesada** [comment]
Guys, just for the record:
these Pull Request MUST NOT be merged after v3.1-rc0.

Only bug fixes.
This pull request must be rollback back and merged for v3.2

The window to merge new features for v3.1 was closed on Friday.

This week is for testing only, and bug fixing.

**Comment by minggo** [comment]
No matter, it will not cause more issues. I have checked out it.

**Comment by zilongshanren** [comment]
Hi Riq, I only had renamed some function names and added some interfaces to keep the code loose coupled.
This PR has many CONFLICTS every time I sync to the base branch, it is really a headache for me.

**Comment by ricardoquesada** [comment]
This feature could work Ok, or not. But the window for adding new features was closed because all our resources should be put on testing v3.1-rc0.

For example, this feature adds a new file, so all the builds needs to be tested again. You said you already tested them, but I think that the WP8 build won't compile because it is was not updated.

Also, for v3.1-rc0 I run the leak detector on Mac and Windows to make sure that not leaks were present on our code.
Now I have to run the leak detector again because this feature as added (and I don't have the time to do it).

I like that we are improving the Layout. UILayout needed some refactor. And I'm happy that we are improving it. But I think we should keep improving it. As an example:
- LayoutExecutant is a somewhat confusing name. I think we need a better name.
- And `RelativeLayoutExecutant::doLayout(LayoutProtocol *layout)` is a huge function. I think it should be split in smaller functions, or refactored.

Perhaps there other features that we can improve as well. I think it doesn't make sense to merge a refactor now, and then merge another refactor a few days later.

To summarize:
- I think we need to improve UILayout, and please keep improving it.
- But the timing to merge this PR is not the best one because:
  - We already closed the window to merge this features (and for a reason).
  - Because it needs more polish

thanks.

**Comment by minggo** [comment]
Ok. 
How to rollback it?

**Comment by ricardoquesada** [comment]
@andyque I'm sorry to hear that I had many conflicts. I know what you mean. 
Conflicts happens to me all the time. Sometimes I have a feature in my own branch for more than 2 months, and I have to keep solving conflicts all the time. But I can't merge the feature if it is not ready, and also I can't merge if the window is closed.

Andyque,  please, join the cocos2d-x-devel@googlegroups.com list. In that list we discuss deadlines, when the merge window is open, when it is closed, etc.
So if you are working on a feature, you know when is the last day that you have in order to send the Pull Request.

@minggo 
I think the easiest way is to send a new PR that undoes this PR.

**Comment by zilongshanren** [comment]
@ricardoquesada ok.

**Comment by ricardoquesada** [comment]
@andyque 
thanks for your understanding. please, keep working on it and I'll merge it for v3.2
My suggestion is:
- what other name can we use for `UILayerExecutant` ?
- can you make `RelatvieLayerExecutant::doLayout()` easier to maintain?

thanks.

**Comment by zilongshanren** [comment]
@ricardoquesada 
Sure, thanks.
- I can't figure out any meaningful name for UILayoutExecutant now, any suggestions?
- For the `RelatvieLayerExecutant::doLayout()` I could extract  some code into small functions and give them a meaningful name. If I could find a pattern of the algorithm, I will do some abstraction to reduce duplications.
